sql/src/test/java/org/elasticsearch/action/bulk/BulkRetryCoordinatorTest.java

```diff
@@ -0,0 +1,164 @@
+            }
+            @Override
+            public void onFailure(Throwable e) {
+            }
+        });
+
+        verify(threadPool).schedule(eq(TimeValue.timeValueMillis(1000)),
+                                    eq(ThreadPool.Names.SAME),
+                                    any(Runnable.class));
+    }
+
+    @Test
+    public void testNoPendingOperationsOnFailedExecution() throws Exception {
+        ThreadPool threadPool = mock(ThreadPool.class);
+        BulkRetryCoordinator coordinator = new BulkRetryCoordinator(threadPool);
+
+        TransportShardUpsertActionDelegate executor = new MockShardUpsertActionDelegate() {
+            @Override
+            public void execute(ShardUpsertRequest request, ActionListener<ShardResponse> listener) {
+                listener.onFailure(new InterruptedException("Dummy execution failed"));
+            }
+        };
+
+        final SettableFuture<ShardResponse> future = SettableFuture.create();
+        coordinator.retry(shardRequest(), executor, new ActionListener<ShardResponse>() {
+            @Override
+            public void onResponse(ShardResponse shardResponse) {
+            }
+            @Override
+            public void onFailure(Throwable e) {
+                future.set(null);
+            }
+        });
+
+        ShardResponse response = future.get();
+        assertNull(response);
+        assertEquals(0, coordinator.numPendingOperations());
+    }
+
+    @Test
+    public void testParallelSuccessfulExecution() throws Exception {
+        ThreadPool threadPool = mock(ThreadPool.class);
+        final BulkRetryCoordinator coordinator = new BulkRetryCoordinator(threadPool);
+
+        final TransportShardUpsertActionDelegate executor = new MockShardUpsertActionDelegate() {
+            @Override
+            public void execute(ShardUpsertRequest request, ActionListener<ShardResponse> listener) {
+                listener.onResponse(new ShardResponse());
+            }
+        };
+
+        final CountDownLatch latch = new CountDownLatch(1000);
+        ExecutorService executorService = Executors.newFixedThreadPool(10, daemonThreadFactory("DummyThreadPool"));
+        for (int i = 0; i < 1000; i++) {
+            executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    coordinator.retry(shardRequest(), executor, new ActionListener<ShardResponse>() {
+                        @Override
+                        public void onResponse(ShardResponse shardResponse) {
+                            latch.countDown();
+                        }
+                        @Override
+                        public void onFailure(Throwable e) {
+                        }
+                    });
+                }
+            });
+        }
+        latch.await();
+        assertEquals(0, coordinator.numPendingOperations());
+        executorService.shutdown();
```

Comment:
please also call awaitTermination with timeout to avoid thread leak assertions from triggering